### PR TITLE
feat: add attach/detach methods to AbstractClientConnector

### LINKED ISSUE #15: Add attach/detach API for com.vaadin.legacy.server.AbstractClientConnector
Attach/Detach behavior that delegates Flow component lifecycle methods:
<html>
<body>

Return | Method
-- | --
Registration | addAttachListener(ClientConnector.AttachListener listener)Add a listener for connector attach events.
Registration | addDetachListener(ClientConnector.DetachListener listener)Add a listener for connector detach events.
void | attach()Notifies the connector that it is connected to a VaadinSession (and therefore also to a UI).
void | detach()Notifies the connector that it is detached from its VaadinSession.
boolean | isAttached()Checks if the connector is attached to a VaadinSession.

</body>
</html>

### PR DESCRIPTION
**Pull request by DiegoCardoso**
Fixes #15

### CODE REVIEW

File: vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                .onComponentEvent(Mockito.any());
+    }
+
+    @Test
+    public void testAttachDetachListenerShouldBeCalledWhenLegacyParentIsAttachedAndDetached() {
+        Component parent = new LegacyComponent(ElementFactory.createDiv());
+        AbstractClientConnector legacyChild = new LegacyComponent(
+                ElementFactory.createDiv());
+
+        legacyChild.attach();
```
Comment (pleku) [inline vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java:81]: Same thing for all the tests

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    boolean isAttached();
+
+    /**
+     * Interface for listening {@link AttachEvent connector attach events}.
+     *
+     */
+    @FunctionalInterface
+    interface AttachListener
+            extends ComponentEventListener<ClientConnector.AttachEvent> {
+    }
```
Comment (pleku) [inline vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java:82]: This is not actually compatible with the legacy API when not used as an lambda, which is probably the most typical use case. But backwards compatibility for all cases can be achieved by
```suggestion
        @Override
        default void onComponentEvent(ClientConnector.AttachEvent event) {
            this.attach(event);
        }

        /**
         * Called when a AttachListener is notified of a AttachEvent.
         *
         * @param event
         *            The attach event that was fired.
         */
         void attach(ClientConnector.AttachEvent event);
    }
```

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Interface for listening {@link DetachEvent connector attach events}.
+     *
+     */
+    @FunctionalInterface
+    interface DetachListener
+            extends ComponentEventListener<ClientConnector.DetachEvent> {
+    }
```
Comment (pleku) [inline vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java:91]: This is not actually compatible with the legacy API when not used as an lambda, which is probably the most typical use case. But backwards compatibility for all cases can be achieved by
```suggestion
        @Override
        default void onComponentEvent(ClientConnector.DetachEvent event) {
            this.detach(event);
        }

        /**
         * Called when a legacy DetachListener is notified of a DetachEvent.
         *
         * @param event
         *            The detach event that was fired.
         */
         void detach(ClientConnector.DetachEvent event);
    }
```

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Notifies the connector that it is detached from its VaadinSession.
+     *
+     * </p>
+     */
+    void detach();
+
+    /**
+     * Checks if the connector is attached to a VaadinSession.
```
Comment (pleku) [inline vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java:69]: Would be great to clarify on `AbstractClientConnector` level method javadocs that this method is delegating to Flow's `Component::isAttached`

**Review by pleku — CHANGES_REQUESTED** [review]
Further API parity can be achieved with a small tweak.

**Review by DiegoCardoso — COMMENTED** [review]
I addressed the comments on the review. Now I can add the methods added by #26 on `ClientConnector` interface.

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                                 *
+                                 */
+
+import com.vaadin.flow.component.Component;
+import com.vaadin.flow.component.ComponentEvent;
+import com.vaadin.flow.component.ComponentEventListener;
+import com.vaadin.flow.shared.Registration;
+import com.vaadin.flow.component.ComponentUtil;
+
+public interface ClientConnector {
```
[the remark this answers is not in the document]
Comment (DiegoCardoso) [reply vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java:36]: Done.

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java (hunk 4 above)
Comment (DiegoCardoso) [reply vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java:80]: Added to `AbstractClientConnector::isAttached`. Not sure if it's clear enough, though.

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java (hunk 3 above)
Comment (DiegoCardoso) [reply vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java:128]: Done.

File: vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    @Test
+    public void attachListenerShouldBeCalled() {
+        AbstractClientConnector connector = Mockito.mock(
+                AbstractClientConnector.class, Mockito.CALLS_REAL_METHODS);
+
+        ClientConnector.AttachListener listener = Mockito
+                .mock(ClientConnector.AttachListener.class);
+        connector.addAttachListener(listener);
+
+        connector.attach();
```
[the remark this answers is not in the document]
Comment (DiegoCardoso) [reply vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java]: Changed to `ComponentUtil::onComponentAttach`.

File: vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    @Test
+    public void detachListenerShouldBeCalled() {
+        AbstractClientConnector connector = Mockito.mock(
+                AbstractClientConnector.class, Mockito.CALLS_REAL_METHODS);
+
+        ClientConnector.DetachListener listener = Mockito
+                .mock(ClientConnector.DetachListener.class);
+        connector.addDetachListener(listener);
+
+        connector.detach();
```
[the remark this answers is not in the document]
Comment (DiegoCardoso) [reply vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java]: Changed to `ComponentUtil::onComponentDetach`.

File: vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java (hunk 1 above)
Comment (DiegoCardoso) [reply vaadin-legacy-components-flow/src/test/java/com/vaadin/legacy/server/AbstractClientConnectorTest.java]: Done.

**Review by pleku — APPROVED** [review]
LGTM
